
<!DOCTYPE HTML>
<html lang="zh-hans" >
    <head>
        <meta charset="UTF-8">
        <meta content="text/html; charset=utf-8" http-equiv="Content-Type">
        <title>Chapter 2 Import data · Bingdong Liu</title>
        <meta http-equiv="X-UA-Compatible" content="IE=edge" />
        <meta name="description" content="">
        <meta name="generator" content="GitBook 3.2.3">
        <meta name="author" content="Bingdong Liu">
        
        
    
    <link rel="stylesheet" href="../gitbook/style.css">

    
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-code/plugin.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-pageview-count/plugin.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-tbfed-pagefooter/footer.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-prism/prism-tomorrow.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-search-pro/search.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-expandable-chapters/expandable-chapters.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-back-to-top-button/plugin.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-splitter/splitter.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-auto-scroll-table/website.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-chapter-fold/chapter-fold.css">
                
            
                
                <link rel="stylesheet" href="../gitbook/gitbook-plugin-fontsettings/website.css">
                
            
        

    

    
        
    

        
    
    
    
    <meta name="HandheldFriendly" content="true"/>
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <meta name="apple-mobile-web-app-capable" content="yes">
    <meta name="apple-mobile-web-app-status-bar-style" content="black">
    <link rel="apple-touch-icon-precomposed" sizes="152x152" href="../gitbook/images/apple-touch-icon-precomposed-152.png">
    <link rel="shortcut icon" href="../gitbook/images/favicon.ico" type="image/x-icon">

    
    <link rel="next" href="2 创建MAE对象.html" />
    
    
    <link rel="prev" href="1 EasyMultiProfiler初识.html" />
    

    
        <link rel="shortcut icon" href='../assets/favicon.ico' type="image/x-icon">
    
    
        <link rel="bookmark" href='../assets/favicon.ico' type="image/x-icon">
    
    
        <link rel="apple-touch-icon" href='../assets/favicon.ico'>
    
    
        
        <link rel="apple-touch-icon" sizes="120x120" href="../assets/favicon.ico">
        
        <link rel="apple-touch-icon" sizes="180x180" href="../assets/favicon.ico">
        
    

    </head>
    <body>
        
<div class="book">
    <div class="book-summary">
        
            
<div id="book-search-input" role="search">
    <input type="text" placeholder="输入并搜索" />
</div>

            
                <nav role="navigation">
                


<ul class="summary">
    
    
    
        
        <li>
            <a href="https://github.com/liubingdong/EasyMultiProfiler" target="_blank" class="custom-link">EasyMultiProfiler</a>
        </li>
    
    

    
    <li class="divider"></li>
    

    
        
        
    
        <li class="chapter " data-level="1.1" data-path="../">
            
                <a href="../">
            
                    
                    Tutorial
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.1.1" data-path="1 EasyMultiProfiler初识.html">
            
                <a href="1 EasyMultiProfiler初识.html">
            
                    
                    Chapter 1 Introduction
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.1.1.1" data-path="1 EasyMultiProfiler初识.html">
            
                <a href="1 EasyMultiProfiler初识.html#11">
            
                    
                    1.1 Installation and update
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.1.1.2" data-path="1 EasyMultiProfiler初识.html">
            
                <a href="1 EasyMultiProfiler初识.html#12">
            
                    
                    1.2 basic information
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter active" data-level="1.1.2" data-path="2 创建MAE对象.html">
            
                <a href="2 创建MAE对象.html">
            
                    
                    Chapter 2 Import data
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.1.2.1" data-path="2 创建MAE对象.html">
            
                <a href="2 创建MAE对象.html#21">
            
                    
                    2.1 Multi-omics data
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.1.2.2" data-path="2 创建MAE对象.html">
            
                <a href="2 创建MAE对象.html#22">
            
                    
                    2.2 Single-omics data
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.1.2.3" data-path="2 创建MAE对象.html">
            
                <a href="2 创建MAE对象.html#23">
            
                    
                    2.3 Save and load data
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.1.3" data-path="3 EasyMultiProfiler内置示例MAE对象.html">
            
                <a href="3 EasyMultiProfiler内置示例MAE对象.html">
            
                    
                    Chapter 3 Demo data
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.1.3.1" data-path="3 EasyMultiProfiler内置示例MAE对象.html">
            
                <a href="3 EasyMultiProfiler内置示例MAE对象.html#31">
            
                    
                    3.1 Load MAE 
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.1.3.2" data-path="3 EasyMultiProfiler内置示例MAE对象.html">
            
                <a href="3 EasyMultiProfiler内置示例MAE对象.html#32">
            
                    
                    3.2 Summary MAE
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.1.4" data-path="4 数据提取 Data Extract.html">
            
                <a href="4 数据提取 Data Extract.html">
            
                    
                    Chapter 4 Data Extract
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.1.4.1" data-path="4 数据提取 Data Extract.html">
            
                <a href="4 数据提取 Data Extract.html#41">
            
                    
                    4.1 EMP_assay_extract
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.1.4.2" data-path="4 数据提取 Data Extract.html">
            
                <a href="4 数据提取 Data Extract.html#42">
            
                    
                    4.2 EMP_rowdata_extract
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.1.4.3" data-path="4 数据提取 Data Extract.html">
            
                <a href="4 数据提取 Data Extract.html#43">
            
                    
                    4.3 EMP_coldata_extract
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.1.5" data-path="5 数据预处理 Data Preparetion.html">
            
                <a href="5 数据预处理 Data Preparetion.html">
            
                    
                    Chapter 5 Data Preparation
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.1.5.1" data-path="5.1 数据折叠 EMP_collapse.html">
            
                <a href="5.1 数据折叠 EMP_collapse.html">
            
                    
                    5.1 EMP_collapse
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.1.5.2" data-path="5.2 批次校正 EMP_adjust_abudance.html">
            
                <a href="5.2 批次校正 EMP_adjust_abudance.html">
            
                    
                    5.2 EMP_adjust_abudance
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.1.5.3" data-path="5.3 实验数据的特征转换 EMP_feature_convert.html">
            
                <a href="5.3 实验数据的特征转换 EMP_feature_convert.html">
            
                    
                    5.3 EMP_feature_convert
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.1.5.4" data-path="5.4 数据标准化 EMP_decostand.html">
            
                <a href="5.4 数据标准化 EMP_decostand.html">
            
                    
                    5.4 EMP_decostand
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.1.5.5" data-path="5.5 数据插补 EMP_impute.html">
            
                <a href="5.5 数据插补 EMP_impute.html">
            
                    
                    5.5 EMP_impute
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.1.5.6" data-path="5.6 确定核心数据 EMP_identify_assay.html">
            
                <a href="5.6 确定核心数据 EMP_identify_assay.html">
            
                    
                    5.6 EMP_identify_assay
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.1.5.7" data-path="5.7 调整数据 EMP_modify_assay.html">
            
                <a href="5.7 调整数据 EMP_modify_assay.html">
            
                    
                    5.7 EMP_modify_assay
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.1.5.8" data-path="5.8 数据重抽样 EMP_rrarefy.html">
            
                <a href="5.8 数据重抽样 EMP_rrarefy.html">
            
                    
                    5.8 EMP_rrarefy
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.1.6" data-path="6 数据分析 Data Analysis.html">
            
                <a href="6 数据分析 Data Analysis.html">
            
                    
                    Chapter 6 Data Analysis
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.1.6.1" data-path="6.1 α多样性分析 EMP_alpha_analysis.html">
            
                <a href="6.1 α多样性分析 EMP_alpha_analysis.html">
            
                    
                    6.1 EMP_alpha_analysis
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.1.6.2" data-path="6.2 聚类分析 EMP_cluster_analysis.html">
            
                <a href="6.2 聚类分析 EMP_cluster_analysis.html">
            
                    
                    6.2 EMP_cluster_analysis
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.1.6.3" data-path="6.3 相关性分析 EMP_cor_analysis.html">
            
                <a href="6.3 相关性分析 EMP_cor_analysis.html">
            
                    
                    6.3 EMP_cor_analysis
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.1.6.4" data-path="6.4 差异性分析 EMP_diff_analysis.html">
            
                <a href="6.4 差异性分析 EMP_diff_analysis.html">
            
                    
                    6.4 EMP_diff_analysis
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.1.6.5" data-path="6.5 降维分析 EMP_dimension_analysis.html">
            
                <a href="6.5 降维分析 EMP_dimension_analysis.html">
            
                    
                    6.5 EMP_dimension_analysis
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.1.6.6" data-path="6.6 富集分析 EMP_enrich_analysis.html">
            
                <a href="6.6 富集分析 EMP_enrich_analysis.html">
            
                    
                    6.6 EMP_enrich_analysis
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.1.6.7" data-path="6.7 标志物分析 EMP_marker_analysis.html">
            
                <a href="6.7 标志物分析 EMP_marker_analysis.html">
            
                    
                    6.7 EMP_marker_analysis
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.1.6.8" data-path="6.8 GSEA富集分析 EMP_GSEA_analysis.html">
            
                <a href="6.8 GSEA富集分析 EMP_GSEA_analysis.html">
            
                    
                    6.8 EMP_GSEA_analysis
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.1.6.9" data-path="6.9 EMP_WGCNA_cluster_analysis.html">
            
                <a href="6.9 EMP_WGCNA_cluster_analysis.html">
            
                    
                    6.9 EMP_WGCNA_cluster_analysis
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.1.6.10" data-path="6.10 EMP_WGCNA_cor_analysis.html">
            
                <a href="6.10 EMP_WGCNA_cor_analysis.html">
            
                    
                    6.10 EMP_WGCNA_cor_analysis
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.1.6.11" data-path="6.11 联合分析 EMP_multi_analysis.html">
            
                <a href="6.11 联合分析 EMP_multi_analysis.html">
            
                    
                    6.11 EMP_multi_analysis
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.1.7" data-path="7 data support.html">
            
                <a href="7 data support.html">
            
                    
                    Chapter 7 Data support
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.1.7.1" data-path="7.1 数据筛选 data_filter.html">
            
                <a href="7.1 数据筛选 data_filter.html">
            
                    
                    7.1 EMP_filter
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.1.7.2" data-path="7.2 数据历史 EMP_history.html">
            
                <a href="7.2 数据历史 EMP_history.html">
            
                    
                    7.2 EMP_history
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.1.7.3" data-path="7.3 数据结果提取和注入 EMP_result.html">
            
                <a href="7.3 数据结果提取和注入 EMP_result.html">
            
                    
                    7.3 EMP_result
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.1.7.4" data-path="7.4 数据概览 EMP_summary.html">
            
                <a href="7.4 数据概览 EMP_summary.html">
            
                    
                    7.4 EMP_summary
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.1.7.5" data-path="7.5 物种数据导入 EMP_taxonomy_import.html">
            
                <a href="7.5 物种数据导入 EMP_taxonomy_import.html">
            
                    
                    7.5 EMP_taxonomy_import
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.1.7.6" data-path="7.6 功能数据导入 EMP_function_import.html">
            
                <a href="7.6 功能数据导入 EMP_function_import.html">
            
                    
                    7.6 EMP_function_import
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.1.7.7" data-path="7.7 标准数据导入 EMP_normal_import.html">
            
                <a href="7.7 标准数据导入 EMP_normal_import.html">
            
                    
                    7.7 EMP_normal_import
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.1.7.8" data-path="7.8 快捷数据导入 EMP_easy_import.html">
            
                <a href="7.8 快捷数据导入 EMP_easy_import.html">
            
                    
                    7.8 EMP_easy_import
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.1.8" data-path="8 data virtualization.html">
            
                <a href="8 data virtualization.html">
            
                    
                    Chapter 8 Data Virtualization
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.1.8.1" data-path="8.1 箱型图 EMP_boxplot.html">
            
                <a href="8.1 箱型图 EMP_boxplot.html">
            
                    
                    8.1 EMP_boxplot
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.1.8.2" data-path="8.2 散点图 EMP_scatterplot.html">
            
                <a href="8.2 散点图 EMP_scatterplot.html">
            
                    
                    8.2 EMP_scatterplot
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.1.8.3" data-path="8.3 富集点状图 EMP_dotplot.html">
            
                <a href="8.3 富集点状图 EMP_dotplot.html">
            
                    
                    8.3 EMP_dotplot
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.1.8.4" data-path="8.4 富集网络图 EMP_netplot.html">
            
                <a href="8.4 富集网络图 EMP_netplot.html">
            
                    
                    8.4 EMP_netplot
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.1.8.5" data-path="8.5 双变量拟合图 EMP_fitline_plot.html">
            
                <a href="8.5 双变量拟合图 EMP_fitline_plot.html">
            
                    
                    8.5 EMP_fitline_plot
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.1.8.6" data-path="8.6 富集曲线图 EMP_curveplot.html">
            
                <a href="8.6 富集曲线图 EMP_curveplot.html">
            
                    
                    8.6 EMP_curveplot
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.1.8.7" data-path="8.7 热图 EMP_heatmap_plot.html">
            
                <a href="8.7 热图 EMP_heatmap_plot.html">
            
                    
                    8.7EMP_heatmap_plot
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.1.8.8" data-path="8.8 桑基图 EMP_sankey_plot.html">
            
                <a href="8.8 桑基图 EMP_sankey_plot.html">
            
                    
                    8.8 EMP_sankey_plot
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.1.8.9" data-path="8.9 结构柱状图 EMP_structure_plot.html">
            
                <a href="8.9 结构柱状图 EMP_structure_plot.html">
            
                    
                    8.9 EMP_structure_plot
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.1.8.10" data-path="8.10 火山图 EMP_volcanol_plot.html">
            
                <a href="8.10 火山图 EMP_volcanol_plot.html">
            
                    
                    8.10 EMP_volcanol_plot
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.1.9" data-path="9 更多有用函数.html">
            
                <a href="9 更多有用函数.html">
            
                    
                    Chapter 9 More useful tools
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.1.9.1" data-path="9 更多有用函数.html">
            
                <a href="9 更多有用函数.html#91">
            
                    
                    9.1 CorRcpp
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.1.9.2" data-path="9 更多有用函数.html">
            
                <a href="9 更多有用函数.html#92">
            
                    
                    9.2 str_detect_multi
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.1.9.3" data-path="9 更多有用函数.html">
            
                <a href="9 更多有用函数.html#93">
            
                    
                    9.3 EMP_to_EMP1
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.1.10" data-path="10 示例教程.html">
            
                <a href="10 示例教程.html">
            
                    
                    Chapter 10 More Examples
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.1.10.1" data-path="10.1 表型数据重分组.html">
            
                <a href="10.1 表型数据重分组.html">
            
                    
                    10.1 Meta data modified
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.1.10.2" data-path="10.2 实验数据快速处理.html">
            
                <a href="10.2 实验数据快速处理.html">
            
                    
                    10.2 Experiment data
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.1.10.3" data-path="10.3 配对队列的处理方案.html">
            
                <a href="10.3 配对队列的处理方案.html">
            
                    
                    10.3 Paired test and intervention corhort
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.2" data-path="Contact.html">
            
                <a href="Contact.html">
            
                    
                    Contact
            
                </a>
            

            
        </li>
    

    

    <li class="divider"></li>

    <li>
        <a href="https://www.gitbook.com" target="blank" class="gitbook-link">
            本书使用 GitBook 发布
        </a>
    </li>
</ul>


                </nav>
            
        
    </div>

    <div class="book-body">
        
            <div class="body-inner">
                
                    

<div class="book-header" role="navigation">
    

    <!-- Title -->
    <h1>
        <i class="fa fa-circle-o-notch fa-spin"></i>
        <a href=".." >Chapter 2 Import data</a>
    </h1>
</div>




                    <div class="page-wrapper" tabindex="-1" role="main">
                        <div class="page-inner">
                            
<div id="book-search-results">
    <div class="search-noresults">
    
                                <section class="normal markdown-section">
                                
                                <h2 id="2-create-multiassayexperiment-object">2. Create MultiAssayExperiment Object</h2>
<p>This section illustrates how to import data from multi-omics projects into the analysis process.</p>
<h3 id="21">2.1 Multi-omics Data</h3>

<p>When the user processes data from multi-omics projects, the data needs to be imported and combined into a <code>MAE</code> object.</p>
<div style="text-align: center;">     
  <img src="../tutorial_figs/EMP_structure1.jpg" style="zoom: 35%;"> 
</div>

<h4 id="211-create-a-relationship-table-of-sample-id-and-subject-id">2.1.1 Create a relationship table of sample ID and subject ID</h4>
<p>In the study, each subject has a unique subject ID (e.g., P1, P2, P3...), and each subject may collect a variety of biological samples and measure corresponding data, including species annotation (taxonomy) for 16s rRNA gene sequencing, KO annotation (geno_ko) for microbial functional genes, eczymatic annotation (geno_ec) for microbial functional genes, etc. When analyzing data from one or more projects, in most cases the subject ID and sample ID are the same (e.g., subject ID P1, project name taxonomy,  sample ID P1). But sometimes we also face the following situations:</p>
<ul>
<li><strong>Only single-omic project</strong>: the subject ID and the sample ID are inconsistent (e.g., subject ID P1, project name taxonomy, sample ID tax_1);</li>
<li><strong>Multi-omics projects</strong>: Within the same subject, sample ID of different projects are inconsistent (e.g., subject ID P1; project name taxonomy, sample ID tax_1; project name geno_ko, sample ID genomics_1).</li>
</ul>
<p>Therefore, in order to effectively manage and analyse this data, users need to create a relationship table (usually named <code>DFMAP</code>, users can also customize) to match the subject ID to the sample ID. Based on the relationship table, users are able to easily import all the data into the <code>MultiAssayExperiment</code> object. During subsequent analysis, the <code>EasyMultiProfiler</code> package automatically replaces the sample ID the subject ID according to the relationship table.</p>
<p>The relationship table-based analysis process has the following benefits:</p>
<ul>
<li><strong>Data Integration and Consistency</strong>: Relationship table ensure that the sample information in the datasets of different projects is properly mapped and integrated, thus ensuring data consistency.</li>
<li><strong>Automated processing</strong>: Data can be managed and processed automatically without manual modifications, so you don&apos;t have to worry about inconsistent naming.</li>
<li><strong>Streamlined analysis workflow</strong>: Consolidating data and automatically replacing sample ID with subject ID makes the analysis process smoother and clearer, reducing errors that can result from manual processing.</li>
</ul>
<div style="background-color: #ffffcc; border: 1px solid #ffcc00; padding: 10px;">
Note&#xFF1A;<br>Do not confuse &quot;Project Name&quot;, &quot;Subject ID&quot;, and &quot;Sample ID&quot;.
</div>

<div style="text-align: center;">     
  <img src="../tutorial_figs/EMP_dfmap_demo.jpg" style="zoom: 50%;"> 
</div>

<p>&#x1F3F7;&#xFE0F;<strong>Example:</strong> Create a relationship table of subject ID and sample ID.</p>
<div style="text-align: center;">     
  <img src="../tutorial_figs/dfmap.jpg" style="zoom: 50%;"> 
</div>

<p>This table needs to be saved as a file and read into R. Users can also manually create data frame directly within R.</p>
<pre class="language-"><code class="lang-R">dfmap &lt;- read.table(&apos;dfmap.txt&apos;,header=TRUE,sep=&apos;\t&apos;)
</code></pre>
<div style="background-color: #ffffcc; border: 1px solid #ffcc00; padding: 10px;">
Note&#xFF1A;<br>
&#x2460;The relationship table must consist of the following three columns: assay (project name), primary (subject ID), colname (sample ID).<br>
&#x2461;The sample ID (e.g., tax_1, genomics_1, etc. in this tutorial) is generated by the upstream annotation.<br>
&#x2462;The sample ID is allowed to be same as the subject ID (e.g., subject ID P1, project name taxonomy, sample ID P1).<br>
&#x2463;The sample ID (or the subject ID corresponding to the sample ID) of each project must be a subset (or identical) of the subject ID of the sample-related data. If the sample ID in the project do not correspond to the subject ID in the sample-related data, these sample ID should be excluded from these projects.<br>
&#x2464;The sample ID of each omics item must correspond to the colname of the relationship table one by one.<br>
&#x2465;For the numbering scheme of project data before and after intervention in the same group of subjects, please refer to section 10.3.
</div>

<h4 id="212-precautions-for-reading-sample-related-data">2.1.2 Precautions for reading sample-related data</h4>
<p>Sample-related data refers to the basic information of the subject, usually including: the subject&apos;s general demographic data (gender, age, residence, etc.), anthropometric data (height, weight, waist circumference, etc.), psychological scale score results (anxiety scale, depression scale, sleep scale, etc.), test results (blood routine, liver function, inflammatory factors, etc.).</p>
<div style="background-color: #ffffcc; border: 1px solid #ffcc00; padding: 10px;">
Note:<br>
&#x2460; If there are missing values in the sample-related data, please leave them blank directly without using placeholders like NA, -, missing, etc.<br>
&#x2461; Please set the first column of the sample-related data as the Sample ID, and set it as the row names when reading (important).<br>
&#x2462; Sample ID must be unique and cannot be duplicated.<br>
&#x2463; It is recommended to set the values of categorical variables in the data as strings rather than pure numbers to avoid unnecessary issues.<br>
&#x2464; For the numbering scheme of project data before and after intervention for the same group of subjects, please refer to Chapter 10.3.
</div>

<pre class="language-"><code class="lang-R">## row.names=1 
meta_data &lt;- read.table(&apos;col.txt&apos;,header = T,row.names = 1)
</code></pre>
<p><img src="../tutorial_figs/meta_data.jpg" alt=""></p>
<h4 id="213-read-microbiome-data">2.1.3 Read microbiome data</h4>
<p>Microbiome data refers to the result files containing the abundance information of microorganisms in each sample, which are generated by various upstream microbial annotation tools. Common result files include the following five types: ASV/OTU tables, abundance tables (kingdom, phylum, class, order, family, genus and species-level), generic-level abundance tables generated by MetaPhlan and Humann, Biom files, and QZV files specific to QIIME2. The module <code>EMP_taxonomy_import</code> supports the input of result files in all of the above formats and converts them into <code>SummariseExperiment</code> objects.</p>
<h5 id="2131-asvotu-forms">2.1.3.1 ASV/OTU forms</h5>
<p>The <code>#</code> characters on the first and second lines of this file do not need to be modified. The data file must contains species annotation information.</p>
<div style="background-color: #ffffcc; border: 1px solid #ffcc00; padding: 10px;">
Note:<br>
&#x2460; The data file must include a column named <code>taxonomy</code>.<br>
&#x2461; The <code>taxonomy</code> column must contain microbial annotation.<br>
&#x2462; It is recommended to use <code>;</code> character to separate microbial annotations.
</div>

<p><img src="../tutorial_figs/OTU.jpg" alt=""></p>
<pre class="language-"><code class="lang-R">tax_data &lt;- EMP_taxonomy_import(&apos;tax.txt&apos;,duplicate_feature=TRUE)
</code></pre>
<h5 id="2132-abundance-table-kingdom-phylum-family-genus-and-species-level">2.1.3.2 Abundance table (Kingdom, Phylum, family, genus, and species-level)</h5>
<p>Since high-taxonomic data can be converted to low-taxonomic data (e.g., Species data can be converted to Phylum data), it is recommended that users directly import data at the highest taxonomy level.</p>
<div style="background-color: #ffffcc; border: 1px solid #ffcc00; padding: 10px;">
Note:<br>
Module <code>EMP_taxonomy_import</code> has a parameter <code>assay_name</code> which defaults to counts (absolute abundance). If the actual data is in relative abundance, you can input relative.
</div>

<p><img src="../tutorial_figs/level-7.jpg" alt=""></p>
<pre class="language-"><code class="lang-R">tax_data &lt;- EMP_taxonomy_import(&apos;tax.txt&apos;)
</code></pre>
<h5 id="2133-classification-annotation-table-for-metaphlan-and-humann">2.1.3.3 Classification annotation table for MetaPhlan and Humann</h5>
<p>The species annotations of this annotation table have a hierarchical structure and usually contain all the result information at each level of the genus and species of the phylum family. The module <code>EMP_taxonomy_import</code> can automatically identify annotation information at the highest classification level and import the complete classification data.</p>
<div style="background-color: #ffffcc; border: 1px solid #ffcc00; padding: 10px;">
Note:<br>
When the input data is in this format, users need to specify the parameter <code>humann_format=TRUE</code> within the module <code>EMP_taxonomy_import</code>.
</div>

<p><img src="../tutorial_figs/humann_tax.jpg" alt=""></p>
<pre class="language-"><code class="lang-R">tax_data &lt;- EMP_taxonomy_import(&apos;tax.txt&apos;,humann_format=TRUE)
</code></pre>
<h5 id="2134-biom-format-files">2.1.3.4 Biom format files</h5>
<p>The biom format is a common species annotation result file in the QIIME1 process, and some users will also save the species annotation file of the QIIME2 process in biom format for data storage. The module <code>EMP_taxonomy_import</code> can directly read the biom format file and import the data information.</p>
<div style="background-color: #ffffcc; border: 1px solid #ffcc00; padding: 10px;">
Note:<br>
When the input data is in this format, users need to specify the parameter <code>file_format=&apos;biom&apos;</code> within the module <code>EMP_taxonomy_import</code>.
</div>

<pre class="language-"><code class="lang-R">tax_data &lt;- EMP_taxonomy_import(&apos;tax.biom&apos;,
                                file_format=&apos;biom&apos;,duplicate_feature=TRUE)
</code></pre>
<h5 id="2135-tabular-files-converted-from-biom-format-files">2.1.3.5 Tabular files converted from Biom format files</h5>
<p>Biom files can be converted into tabular files using <a href="https://biom-format.org/documentation/biom_conversion.html" target="_blank">the Biom Convert method</a>, or imported directly.</p>
<div style="background-color: #ffffcc; border: 1px solid #ffcc00; padding: 10px;">
&#x6CE8;&#x610F;&#xFF1A;<br>
When the input data is in this format, users do not need to modify the header and # character.
</div>

<pre class="language-"><code class="lang-R">tax_data &lt;- EMP_taxonomy_import(&apos;tax.txt&apos;,duplicate_feature=TRUE)
</code></pre>
<p><img src="../tutorial_figs/out_table.jpg" style="zoom:67%;"></p>
<h5 id="2136-qzv-format-file">2.1.3.6 QZV format file</h5>
<p>In the Qiime2 process, the qiime taxa barplot generates a QZV file of microbial annotation results. The module <code>EMP_taxonomy_import</code> can directly read the QZV format file and import the data information.</p>
<div style="background-color: #ffffcc; border: 1px solid #ffcc00; padding: 10px;">
Note:<br>
When the input data is in this format, users need to specify the parameter <code>file_format=&apos;qzv&apos;</code> within the module <code>EMP_taxonomy_import</code>.
</div>

<pre class="language-"><code class="lang-R">tax_data &lt;- EMP_taxonomy_import(&apos;tax.qzv&apos;,
                                file_format=&apos;qzv&apos;,duplicate_feature=TRUE)
</code></pre>
<h4 id="214-read-koec-omics-data">2.1.4 Read KO/EC omics data</h4>
<h5 id="2141-ordinary-tabular-form">2.1.4.1 Ordinary tabular form</h5>
<p>KO/EC omics project data in common format refers to metagenomic or metatranscriptomic annotated result files annotated with KO or EC ID. The module <code>EMP_function_import</code> can directly read the annotation table and query the corresponding annotation information integrated by the KEGG database.</p>
<div style="background-color: #ffffcc; border: 1px solid #ffcc00; padding: 10px;">
Note:<br>
&#x2460; In the encoding, use only pure KO and EC ID, for example: instead of KO:K00010, use only K00010; instead of ec:1.1.1.1, use only 1.1.1.1.<br>
&#x2461; Within the module <code>EMP_function_import</code>, the parameter <code>assay_name</code> defaults to counts (absolute abundance). Users should input TPM, FPKM, etc., based on the actual data.
</div>

<center class="half">    
  <img src="../tutorial_figs/input_ko.jpg" style="zoom:22%;">  
  <img src="../tutorial_figs/input_ec.jpg" style="zoom:22%;">
</center>

<pre class="language-"><code class="lang-R">ko_data &lt;- EMP_function_import(&apos;ko.txt&apos;,type = &apos;ko&apos;)
ec_data &lt;- EMP_function_import(&apos;ec.txt&apos;,type = &apos;ec&apos;)
</code></pre>
<h5 id="2141-humann-format">2.1.4.1 Humann format</h5>
<p>KO/EC omics project data in Humann format refers to the hierarchical KO/EC annotation tables generated by the Humann tool. The module <code>EMP_function_import</code> ignores the information of UNGROUPED, directly reads the comment table, and queries the KEGG database to integrate the corresponding comment information.</p>
<div style="background-color: #ffffcc; border: 1px solid #ffcc00; padding: 10px;">
&#x6CE8;&#x610F;&#xFF1A;<br>
  The header of the file needs to be cleaned up and characters such as # removed.
</div>

<p><img src="../tutorial_figs/humann_ko.jpg" alt=""></p>
<pre class="language-"><code class="lang-R">ko_data &lt;- EMP_function_import(&apos;ko.txt&apos;,type = &apos;ko&apos;,humann_format = TRUE)
</code></pre>
<h4 id="215-read-standard-table-information">2.1.5 Read standard table information</h4>
<p>Standard tabular data usually refers to data such as transcriptome or metabolome, which has no special format and can be imported directly.</p>
<h5 id="2151-data-import-without-feature-annotation-information">2.1.5.1 Data import without feature annotation information</h5>
<p>This data is commonly used in transcriptome or genomic data, with &quot;rows&quot; as features and &quot;columns&quot; as samples.</p>
<p><img src="../tutorial_figs/input_tran.jpg" style="zoom:67%;"></p>
<pre class="language-"><code class="lang-R">tran_data &lt;- EMP_normal_import(&apos;tran.txt&apos;)
</code></pre>
<h5 id="2152-data-import-containing-feature-annotation-information">2.1.5.2 Data import containing feature annotation information</h5>
<p>This data is commonly used in metabolomics data, with &quot;rows&quot; as features and &quot;columns&quot; as sampleID and feature-related annotations. When importing this kind of data, users need to specify a sampleID column.</p>
<div style="background-color: #ffffcc; border: 1px solid #ffcc00; padding: 10px;">
Note:<br>
&#x2460; Users can directly specify the sampleID using parameter<code>sampleID</code>.<br>
&#x2461; Users can directly specify the sampleID using the relationship table.
</div>

<p><img src="../tutorial_figs/input_metabo.jpg" alt=""></p>
<pre class="language-"><code class="lang-R"># specify the sampleID using parameter sampleID
sample_ID &lt;-c(&quot;Control1&quot;, &quot;Control10&quot;, &quot;Control2&quot;, &quot;Control3&quot;, &quot;Control4&quot;, 
        &quot;Control5&quot;, &quot;Control6&quot;, &quot;Control7&quot;, &quot;Control8&quot;, &quot;Control9&quot;,
        &quot;Treat10&quot;, &quot;Treat1&quot;, &quot;Treat2&quot;, &quot;Treat3&quot;, &quot;Treat4&quot;, &quot;Treat5&quot;, 
        &quot;Treat6&quot;, &quot;Treat7&quot;, &quot;Treat8&quot;, &quot;Treat9&quot;)
metbol_data &lt;- EMP_normal_import(&apos;metabol.txt&apos;,sampleID = sample_ID)

# specify the sampleID using the relationship table
metbol_data &lt;- EMP_normal_import(&apos;metabol.txt&apos;,sampleID = sample_ID,
                                 dfmap = dfmap,assay = &apos;untarget_metabol&apos;)
</code></pre>
<h4 id="216--integrate-all-data-into-the-multiassayexperiment-object">2.1.6  Integrate all data into the MultiAssayExperiment object</h4>
<div style="background-color: #ffffcc; border: 1px solid #ffcc00; padding: 10px;">
Note:<br>
&#x2460; In this example, <code>dfmap</code> refers to the relationship table in section 2.1.1.<br>
&#x2461; The <code>assay</code> column in the relationship table must be converted to a factor.<br>
&#x2462; Names in <code>objlist</code> must match the project names in the <code>assay</code> column of the relationship table.<br>
&#x2463; If assembly fails, users can check the data using the module <code>MultiAssayExperiment::prepMultiAssay(objlist, meta_data, dfmap)</code>.<br>
&#x2464; For more detailed information about the <code>MultiAssayExperiment</code> package, refe to <a href="https://bioconductor.org/packages/release/bioc/vignettes/MultiAssayExperiment/inst/doc/MultiAssayExperiment.html" target="_blank">this tutorial</a>.
</div>

<pre class="language-"><code class="lang-R">#### The assay column in the relationship table must be converted to a factor
dfmap$assay &lt;- as.factor(dfmap$assay)

#### Names in objlist must match the project names in the assay column of the relationship table
objlist &lt;- list(&quot;taxonomy&quot; = tax_data,
                &quot;geno_ko&quot; = ko_data,
                &quot;geno_ec&quot; = ec_data,
                &quot;untarget_metabol&quot; = metbol_data,
                &quot;host_gene&quot; = geno_data)

MAE &lt;- MultiAssayExperiment::MultiAssayExperiment(experiments=objlist, 
                                                  colData=meta_data, 
                                                  sampleMap=dfmap)
</code></pre>
<h3 id="22">2.2 Quickly create a MultiAssayExperiment object for a single-omic project</h3>

<p>When the users only have a single-omic project data (for example, microbiome project data or transcriptomics project data), the module EMP_easy_import can be used to quickly build a <code>MAE</code> object and directly perform downstream analysis processes.</p>
<div style="background-color: #ffffcc; border: 1px solid #ffcc00; padding: 10px;">
Note:<br>
&#x2460; The parameter <code>type</code> includes four options: <code>tax</code>, <code>ko</code>, <code>ec</code>, and <code>normal</code>. Users should determine which to use based on the context.<br>
&#x2461; For single-omic project where no relationship table is involved, the sampleID in the abundance matrix must match those in the associated sample data.
</div>

<pre class="language-"><code class="lang-R">## row.names=1
meta_data &lt;- read.table(&apos;col.txt&apos;,header = T,row.names = 1)
MAE &lt;- EMP_easy_import(&apos;tran.txt&apos;,coldata = meta_data,type = &apos;normal&apos;)
MAE &lt;- EMP_easy_import(&apos;tax.txt&apos;,coldata = meta_data,type = &apos;tax&apos;)
MAE &lt;- EMP_easy_import(&apos;ko.txt&apos;,coldata = meta_data,type = &apos;ko&apos;)
</code></pre>
<h3 id="23">2.3 MultiAssayExperiment object storage and reading</h3>

<p>Once the <code>MAE</code> object has been created, it can be stored locally for direct reading during the next analysis without having to repeat the creation step.</p>
<div style="background-color: #ffffcc; border: 1px solid #ffcc00; padding: 10px;">
Note&#xFF1A;<br>
Before storing, make sure to set R&apos;s local working directory correctly, and the data files will be stored in this working directory.
</div>

<pre class="language-"><code class="lang-R"># Save in your working diectory
saveRDS(MAE,file = &apos;MAE.rds&apos;)
# Load the object 
MAE &lt;- readRDS(file = &apos;MAE.rds&apos;)
</code></pre>
<footer class="page-footer"><span class="copyright">Copyright &#xA9; 382983280@qq.com 2024 all right reserved&#xFF0C;powered by Gitbook</span><span class="footer-modification">&#x66F4;&#x65B0;&#x65F6;&#x95F4;&#xFF1A;
2024-08-12 21:51:26
</span></footer>
                                
                                </section>
                            
    </div>
    <div class="search-results">
        <div class="has-results">
            
            <h1 class="search-results-title"><span class='search-results-count'></span> results matching "<span class='search-query'></span>"</h1>
            <ul class="search-results-list"></ul>
            
        </div>
        <div class="no-results">
            
            <h1 class="search-results-title">No results matching "<span class='search-query'></span>"</h1>
            
        </div>
    </div>
</div>

                        </div>
                    </div>
                
            </div>

            
                
                <a href="1 EasyMultiProfiler初识.html#12" class="navigation navigation-prev " aria-label="Previous page: 1.2 basic information">
                    <i class="fa fa-angle-left"></i>
                </a>
                
                
                <a href="2 创建MAE对象.html#21" class="navigation navigation-next " aria-label="Next page: 2.1 Multi-omics data">
                    <i class="fa fa-angle-right"></i>
                </a>
                
            
        
    </div>

    <script>
        var gitbook = gitbook || [];
        gitbook.push(function() {
            gitbook.page.hasChanged({"page":{"title":"Chapter 2 Import data","level":"1.1.2","depth":2,"next":{"title":"2.1 Multi-omics data","level":"1.1.2.1","depth":3,"anchor":"#21","path":"MD/2 创建MAE对象.md","ref":"MD/2 创建MAE对象.md#21","articles":[]},"previous":{"title":"1.2 basic information","level":"1.1.1.2","depth":3,"anchor":"#12","path":"MD/1 EasyMultiProfiler初识.md","ref":"MD/1 EasyMultiProfiler初识.md#12","articles":[]},"dir":"ltr"},"config":{"plugins":["code","pageview-count","tbfed-pagefooter","favicon","-highlight","prism","google-tongji-with-multiple-channel","search-pro","expandable-chapters","hide-element","back-to-top-button","splitter","auto-scroll-table","-lunr","-search","-sharing","sharing-plus","chapter-fold","livereload"],"styles":{"website":"assets/styles/website.css"},"pluginsConfig":{"tbfed-pagefooter":{"copyright":"Copyright © 382983280@qq.com 2024","modify_label":"更新时间：","modify_format":"YYYY-MM-DD HH:mm:ss"},"chapter-fold":{},"prism":{"css":["prismjs/themes/prism-tomorrow.css"]},"livereload":{},"splitter":{},"google-tongji-with-multiple-channel":{"multipleChannelConfig":{},"url":"https://www.googletagmanager.com/gtag/js","token":"G-YE9M3Z1JB4"},"search-pro":{},"sharing-plus":{"qq":false,"all":["facebook","google","twitter","instapaper","linkedin","pocket","stumbleupon"],"douban":false,"facebook":true,"weibo":false,"instapaper":false,"whatsapp":false,"hatenaBookmark":false,"twitter":true,"messenger":false,"line":false,"vk":false,"pocket":true,"google":false,"viber":false,"stumbleupon":false,"qzone":false,"linkedin":false},"auto-scroll-table":{},"code":{"copyButtons":true},"hide-element":{"elements":[".gitbook-link"]},"fontsettings":{"theme":"white","family":"sans","size":2},"favicon":{"shortcut":"assets/favicon.ico","bookmark":"assets/favicon.ico","appleTouch":"assets/favicon.ico","appleTouchMore":{"120x120":"assets/favicon.ico","180x180":"assets/favicon.ico"}},"back-to-top-button":{},"pageview-count":{},"sharing":{"all":["facebook","google","linkedin","twitter","weibo","qq"]},"theme-default":{"styles":{"website":"styles/website.css","pdf":"styles/pdf.css","epub":"styles/epub.css","mobi":"styles/mobi.css","ebook":"styles/ebook.css","print":"styles/print.css"},"showLevel":false},"expandable-chapters":{}},"theme":"default","author":"Bingdong Liu","pdf":{"pageBreaksBefore":"/","headerTemplate":null,"paperSize":"a4","margin":{"right":62,"left":62,"top":36,"bottom":36},"fontSize":12,"fontFamily":"Arial","footerTemplate":null,"chapterMark":"pagebreak","pageNumbers":true},"structure":{"langs":"LANGS.md","readme":"README.md","glossary":"GLOSSARY.md","summary":"SUMMARY.md"},"isbn":"","variables":{},"title":"Bingdong Liu","language":"zh-hans","links":{"sharing":{"all":true,"facebook":true,"google":true,"twitter":true,"weibo":true},"sidebar":{"EasyMultiProfiler":"https://github.com/liubingdong/EasyMultiProfiler"}},"gitbook":"*","description":"EasyMultiProfiler.","extension":null},"file":{"path":"MD/2 创建MAE对象.md","mtime":"2024-08-12T13:51:26.000Z","type":"markdown"},"gitbook":{"version":"3.2.3","time":"2024-09-07T14:44:05.421Z"},"basePath":"..","book":{"language":""}});
        });
    </script>
</div>

        
    <script src="../gitbook/gitbook.js"></script>
    <script src="../gitbook/theme.js"></script>
    
        
        <script src="../gitbook/gitbook-plugin-code/plugin.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-pageview-count/plugin.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-google-tongji-with-multiple-channel/plugin.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-search-pro/jquery.mark.min.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-search-pro/search.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-expandable-chapters/expandable-chapters.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-hide-element/plugin.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-back-to-top-button/plugin.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-splitter/splitter.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-auto-scroll-table/plugin.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-sharing-plus/buttons.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-chapter-fold/chapter-fold.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-livereload/plugin.js"></script>
        
    
        
        <script src="../gitbook/gitbook-plugin-fontsettings/fontsettings.js"></script>
        
    

    </body>
</html>

